// cucumber-html-report/lib/template.js
'use strict';

var path = require('path');
var fs = require('fs');

var baseTemplate = function baseTemplate(name) {
  return path.join(__dirname, '..', 'templates', name);
};
var defaultTemplate = baseTemplate('template.mustache');

var loadTemplate = function loadTemplate(templateFile) {
  return fs.readFileSync(templateFile).toString();
};

function getTemplatePartials(partialDirectory) {
  var partials = {};
  var fileList = fs.readdirSync(partialDirectory).map(function (file) {
    if (file[0] === '.') {
      return undefined;
    }

    var fileInfo = file.split('.');
    return {
      name: fileInfo[0],
      extension: fileInfo[1]
    };
  });

  fileList.forEach(function (fileInfo) {
    var templateFile = path.join(partialDirectory, fileInfo.name + '.' + fileInfo.extension);
    partials[fileInfo.name] = loadTemplate(templateFile);
  });

  return partials;
}

exports.defaultTemplate = defaultTemplate;
exports.load = loadTemplate;
exports.getTemplatePartials = getTemplatePartials;
